resources/assets/js/annotations/components/annotationCanvas.vue

```diff
@@ -529,10 +531,37 @@ export default {
                     shape: geometry.getType(),
                     points: this.getPoints(geometry),
                 }, removeCallback);
-            } else {
+            } else if (validDrawing) {
                 this.annotationSource.removeFeature(e.feature);
+            } else {
+                let source = this.annotationSource;
+                source.on('change', (evtChange) => {
+                    if (evtChange.target.getState() === 'ready' && source.hasFeature(e.feature)) {
+                        source.removeFeature(e.feature);
+                        this.$emit('has-invalid-shape', e.feature.getGeometry().getType());
+                    }
+
+                });
```

Comment:
> Also, why emit an event and handle the error message in the container instead of directly in the annotationCanvas?

Because the same was done when a[ label was not selected](https://github.com/biigle/core/blob/67c342464023a29d1c8a74b566ec30836ff0c384/resources/assets/js/annotations/components/annotationCanvas.vue#L557-L564) before annotation. I just copied that for this issue.

Reply:
> This will add a new event listener each time a new (invalid) annotation is created. Why the event listener instead of just calling `removeFeature()`?

Because then you get this error: `TypeError: this.featureChangeKeys_[featureKey] is undefined.`. It has something to do with asynchronous methods which is why the feature is not there yet or anymore. The problem is that the feature is still displayed on the canvas. Only after refreshing the page, the feature disappears.

Reply:
Ok. Then please use `once()` instead of `on()`. Also, the `$emit` could be called immediately (outside of the event handler). Why is the `evtChange.target.getState() === 'ready'` required?

I think you can also remove the `else if (validDrawing)` case above. This would cause the same error and (since the error never happened) this case can/should never happen. Drawing is aborted if no label is selected and the additional check for `hasSelectedLabel` should not be required.

Reply:
Ok. Then please use `once()` instead of `on()`. Also, the `$emit` could be called immediately (outside of the event handler). Why is the `evtChange.target.getState() === 'ready'` required?

I think you can also remove the `else if (validDrawing)` case above. This would cause the same error and (since the error never happened) this case can/should never happen. Drawing is aborted if no label is selected and the additional check for `hasSelectedLabel` should not be required.

Reply:
I got the solution with the event listener from stackoverflow and forgot to check if this condition was necessary. I have removed it.